src/python/pants/backend/docker/target_types.py

```diff
@@ -140,6 +149,46 @@ class DockerSkipPushField(BoolField):
     help = "If set to true, do not push this image to registries when running `./pants publish`."
 
 
+class DockerBuildOptionFieldMixin(ABC):
+    """Inherit this mixin class to provide options to `docker build`."""
+
+    docker_build_option: ClassVar[str]
+
+    @abstractmethod
+    def option_values(self) -> Iterator[str]:
+        """Subclasses must implement this, to turn their `self.value` into none, one or more option
+        values."""
+
+    @final
+    def options(self) -> Iterator[str]:
+        for value in self.option_values():
+            yield f"{self.docker_build_option}={shlex.quote(value)}"
+
+
+class DockerBuildSecretsOptionField(
+    AsyncFieldMixin, DockerBuildOptionFieldMixin, DictStringToStringField
+):
+    alias = "secrets"
+    help = "Secret file to expose to the build (only if BuildKit enabled).\n\n" + dedent(
+        """\
+        Example:
+
+            docker_image(
+                secrets={
+                    "mysecret": "/local/secret",
+                }
+            )
+        """
+    )
+
+    docker_build_option = "--secret"
+
+    def option_values(self) -> Iterator[str]:
+        root = os.path.join(get_buildroot(), self.address.spec_path)
+        for secret, path in (self.value or {}).items():
+            yield f"id={secret},src={os.path.join(root, path)}"
```

Comment:
This is pretty intricate relying on an implementation detail of when you join two absolute paths:

```
>>> os.path.join("/foo", "/var")
'/var'
```

If you stick with that, it could use a comment. Probably better is to use `os.path.isabs` and only join with build root if relevant.